extensions/EXT/EGL_EXT_explicit_device.txt

```diff
@@ -0,0 +1,101 @@
+Name
+
+    EXT_explicit_device
+
+Name Strings
+
+    EGL_EXT_explicit_device
+
+Contributors
+
+    Adam Jackson
+    Nicolai Haehnle
+    Daniel Stone
+
+Contacts
+
+    Adam Jackson <ajax@redhat.com>
+
+Status
+
+    Complete
+
+Version
+
+    Version 2, 2017-08-08
+
+Number
+
+    EGL Extension #122
+
+Extension Type
+
+    EGL device extension
+
+Dependencies
+
+    Requires EGL_EXT_platform_base and EGL_EXT_device_base.
+
+    EGL_EXT_platform_device trivially interacts with this extension.
+
+    This extension is written against the EGL 1.5 Specification.
+
+Overview
+
+    A system may support multiple devices and multiple window systems. For
+    example, a Wayland environment may drive multiple GPUs and support both
+    X11 and EGL clients. In order to realize this, the implementation must
+    allow the client to specify both device and platform explicitly.
+
+    The EGL_EXT_platform_device extension enables the client to create a
+    display from a device, but does so by overloading the native_display
+    argument to eglGetPlatformDisplay; the user passes the device as the
+    native display, and presumably the EGL picks a sensible default for the
+    window system.
+
+    EGL_EXT_explicit_device differs by passing the device as an attribute
+    to eglGetPlatformDisplay. This allows the client to name both the device
+    and the platform.
```

Comment:
I agree this is useful to solve a separate use case from that EGL_EXT_platform_device solves: PRIME/Optimus-style workloads where you want one GPU to power your compositor and another GPU to power a client app, with explicit application choice over the latter.  However, there's no mechanism exposed here to query which devices are compatible with a given native display.  It presumes support.  I think you need an additional function or something to get a list of devices given a native display.

Reply:
@nwnk Yes, your eglQueryDisplayDevicesEXT() API sounds right.  Might want to make it eglQueryNativeDisplayDevicesEXT() to avoid clashing with the various existing EGL display queries that operate on an EGLDisplay object.

Reply:
If you want some way for an application to ask, "What devices could I use with this display?" then you'd need an attribute list as well. With X11 for example, you might get different results on different screens, which are specified using an attribute. So, something like this could work:
```
EGLBoolean eglQueryNativeDisplayDevicesEXT(
        EGLenum platform, void *native_display, const EGLAttrib *attrib_list,
        EGLint max_devices, EGLDeviceEXT *devices, EGLint *num_devices);
```

Another option would be to use eglQueryDevicesEXT to get the device handles, and then add a function to check if a specific device can work with a native display:
```
EGLBoolean eglQueryDisplayDeviceCompatibleEXT(
        EGLenum platform, void *native_display, const EGLAttrib *attrib_list,
        EGLDeviceEXT device);
```

Reply:
Also, for an application to offer something like a drop-down menu for device selection, you'd also need to provide some sort of a persistent identifier for each device that the app could store in a config file. But, that would probably be outside the scope of this extension.